examples/wifi-echo/server/esp32/main/DeviceCallbacks.cpp

```diff
@@ -0,0 +1,182 @@
+        if (esp_netif_get_ip_info(esp_netif_get_handle_from_ifkey("WIFI_STA_DEF"), &ipInfo) == ESP_OK)
+        {
+            char ipAddrStr[INET_ADDRSTRLEN];
+            esp_ip4addr_ntoa(&ipInfo.ip, ipAddrStr, sizeof(ipAddrStr));
+            ESP_LOGI(TAG, "Server ready at: %s:%d", ipAddrStr, CHIP_PORT);
+
+            // Since the commissioner device does not yet have a mechanism to discover the IP address
+            // of the peripheral, the following code send it over the current Rendezvous session.
+            if (rendezvousDelegate != NULL)
+            {
+                rendezvousDelegate->Send(ipAddrStr);
+            }
+        }
+        wifiLED.Set(true);
+    }
+    else if (event->InternetConnectivityChange.IPv4 == kConnectivity_Lost)
+    {
+        ESP_LOGE(TAG, "Lost IPv4 connectivity...");
+        wifiLED.Set(false);
+    }
+    if (event->InternetConnectivityChange.IPv6 == kConnectivity_Established)
+    {
+        ESP_LOGI(TAG, "IPv6 Server ready...");
+    }
+    else if (event->InternetConnectivityChange.IPv6 == kConnectivity_Lost)
+    {
+        ESP_LOGE(TAG, "Lost IPv6 connectivity...");
+    }
+}
+
+void DeviceCallbacks::OnSessionEstablished(const ChipDeviceEvent * event)
+{
+    if (event->SessionEstablished.IsCommissioner)
+    {
+        ESP_LOGI(TAG, "Commissioner detected!");
+    }
+}
+
+void DeviceCallbacks::OnOnOffPostAttributeChangeCallback(uint8_t endpointId, uint16_t attributeId, uint8_t * value)
+{
+    VerifyOrExit(attributeId == ZCL_ON_OFF_ATTRIBUTE_ID, ESP_LOGI(TAG, "Unhandled Attribute ID: '0x%04x", attributeId));
+    VerifyOrExit(endpointId == 1 || endpointId == 2, ESP_LOGE(TAG, "Unexpected EndPoint ID: `0x%02x'", endpointId));
+
+    // At this point we can assume that value points to a bool value.
+    endpointId == 1 ? statusLED1.Set(*value) : statusLED2.Set(*value);
+
+exit:
+    return;
+}
+
+void IdentifyTimerHandler(Layer * systemLayer, void * appState, Error error)
+{
+    statusLED1.Animate();
+
+    // Decrement the timer count.
+    identifyTimerCount--;
+
+    if (identifyTimerCount)
+    {
+        SystemLayer.StartTimer(kIdentifyTimerDelayMS, IdentifyTimerHandler, appState);
+    }
+}
+
+void DeviceCallbacks::OnIdentifyPostAttributeChangeCallback(uint8_t endpointId, uint16_t attributeId, uint8_t * value)
+{
+    VerifyOrExit(attributeId == ZCL_IDENTIFY_TIME_ATTRIBUTE_ID, ESP_LOGI(TAG, "Unhandled Attribute ID: '0x%04x", attributeId));
+    VerifyOrExit(endpointId == 1, ESP_LOGE(TAG, "Unexpected EndPoint ID: `0x%02x'", endpointId));
+
+    statusLED1.Blink(kIdentifyTimerDelayMS * 2);
+
+    // timerCount represents the number of callback execution before the we stopped the timer.
+    // value is expressed in seconds and the timer is fired every 250ms, so just multiply value by 4.
+    // Also, we want timerCount to be odd number, so the ligth state ends in the same state it starts.
+    identifyTimerCount = (*value) * 4;
+
+    SystemLayer.CancelTimer(IdentifyTimerHandler, this);
```

Comment:
nice.  note that the Callback based timer API handles cancellation for you automagically